File: src/main/java/ting/leetcode/java/binarytree/EvenOddTree.java
package ting.leetcode.java.binarytree;

import java.util.ArrayDeque;
import java.util.Queue;

public class EvenOddTree {
    public boolean isEvenOddTree(TreeNode root) {
        if (root == null) return true;
        int step = 0;
        Queue<TreeNode> nodes = new ArrayDeque<>();

        nodes.offer(root);
        while (!nodes.isEmpty()) {
            int size = nodes.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = nodes.poll();

                if ((step & 1) == 1) {
                    if ((node.val & 1) == 1) {
                        System.out.println("f1");
                        return false;
                    }

                    if (!nodes.isEmpty() && i < size - 1) {
                        if (node.val <= nodes.peek().val) {
                            return false;
                        }
                    }
                } else {
                    if ((node.val & 1) == 0) {
                        return false;
                    }

                    if (!nodes.isEmpty() && i < size - 1) {
                        if (node.val >= nodes.peek().val) {
                            return false;
                        }
                    }
                }

                if (node.left != null) {
                    nodes.offer(node.left);
                }

                if (node.right != null) {
                    nodes.offer(node.right);
                }
            }
            step++;
        }
        return true;
    }
}

File: src/main/kotlin/ting/leetcode/tree/EvenOddTree.kt
package ting.leetcode.tree

class EvenOddTree {
    fun isEvenOddTree(root: TreeNode?): Boolean {
        root ?: return true
        var level = 0
        val nodes = ArrayDeque<TreeNode>()

        nodes.add(root)
        while (nodes.isNotEmpty()) {
            val size = nodes.size
            for (i in 0 until size) {
                val node = nodes.removeFirst()

                if ((level and 1) == 1) {
                    if ((node.`val` and 1) == 1) {
                        return false
                    }

                    nodes.firstOrNull()
                        ?.let { if (i < size - 1 && node.`val` <= it.`val`) return false }
                } else {
                    if ((node.`val` and 1) == 0) {
                        return false
                    }

                    nodes.firstOrNull()
                        ?.let { if (i < size - 1 && node.`val` >= it.`val`) return false }
                }

                node.left?.let { nodes.add(it) }
                node.right?.let { nodes.add(it) }
            }
            level++
        }
        return true
    }
}
